lib/src/main/java/dev/openfeature/javasdk/OpenFeatureClient.java

```diff
@@ -122,6 +124,7 @@ private <T> void afterHooks(HookContext hookContext, FlagEvaluationDetails<T> de
     private EvaluationContext beforeHooks(HookContext hookCtx, List<Hook> hooks, ImmutableMap<String, Object> hints) {
         // These traverse backwards from normal.
         EvaluationContext ctx = hookCtx.getCtx();
+        // This is a bit counter-intuitive to me, just want to make sure... this is the stage that gets reversed because ArrayList.addAll() does a prepend, right?
```

Comment:
Before order is: 1,2,3,4
after,error,finally order is: 4,3,2,1

So we maintain 4,3,2,1 order and reverse for before.

Reply:
Right, I understand that, it's just that reading this, it seems that before is running in reverse (4,3,2,1), you are reversing the ArrayList. I'm just confirming that this is because the arraylist already exists in order (4,3,2,1) because items are PRE-pended with addAll(), right?

Reply:
<img width="344" alt="Screen Shot 2022-06-15 at 8 51 14 PM" src="https://user-images.githubusercontent.com/3853/173987425-34be0284-1aa8-46d9-bd62-7a822702c1fc.png">
They aren't prepended via addAll(), but they are prepended via this above.
jdoodle.com/ia/sfr is an example of what addAll does, if it helps.

--

It sounds like you might have a really good test case in your head. Mind writing that out and I'll add it to the suite?

Reply:
I think I'm doing a bad job communicating about this. `addAll()` does seem to "pre-prend" if I understand the [doc](https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/List.html#addAll(int,java.util.Collection)) right.

> Shifts the element currently at that position (if any) and any subsequent elements to the right (increases their indices).

The case I'm wondering about is if you did:

```
client.addHooks(before1, before2)
```
before1 will run before2. As the code is written, it seem like this will happen because of the way `addAll` works.